File: LeetCode/101/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:

        def isMirror(left, right):
            if (left is None and right is None):
                return True
            if (left is None or right is None):
                return False

            if (left.val == right.val):
                inner_pair = isMirror(left.right, right.left)
                outer_pair = isMirror(left.left, right.right)
                return inner_pair and outer_pair
            else:
                return False

        if (root is None):
            return True
        else:
            return isMirror(root.left, root.right)

File: LeetCode/101/main.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

bool check(struct TreeNode* a, struct TreeNode* b) {

    if(a == NULL && b == NULL) {
        return true;
    }
    if(a == NULL || b == NULL) {
        return false;
    }
    if(a->val != b->val) {
        return false;
    }
    return check(a->left, b->right) && check(a->right, b->left);

}

bool isSymmetric(struct TreeNode* root) {
    if (root == NULL) return true;
    return check(root->left, root->right);
}